californium-core/src/main/java/org/eclipse/californium/core/network/stack/TcpObserveLayer.java

```diff
@@ -0,0 +1,70 @@
+/*******************************************************************************
+/*******************************************************************************
+ * Copyright (c) 2015 Bosch Software Innovations GmbH and others.
+ * 
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * and Eclipse Distribution License v1.0 which accompany this distribution.
+ * 
+ * The Eclipse Public License is available at
+ *    http://www.eclipse.org/legal/epl-v10.html
+ * and the Eclipse Distribution License is available at
+ *    http://www.eclipse.org/org/documents/edl-v10.html.
+ * 
+ * Contributors:
+ *    Achim Kraus (Bosch Software Innovations GmbH) - Initial implementation,
+ *                                                    Derived from ObserveLayer
+ ******************************************************************************/
+package org.eclipse.californium.core.network.stack;
+
+import java.util.logging.Logger;
+
+import org.eclipse.californium.core.coap.Request;
+import org.eclipse.californium.core.coap.Response;
+import org.eclipse.californium.core.network.Exchange;
+import org.eclipse.californium.core.network.config.NetworkConfig;
+import org.eclipse.californium.core.observe.ObserveRelation;
+
+public class TcpObserveLayer extends AbstractLayer {
+
+	private static final Logger LOGGER = Logger.getLogger(TcpObserveLayer.class.getName());
+
+	private static final Integer CANCEL = 1;
+
+	public TcpObserveLayer(NetworkConfig config) {
+		// so far no configuration values for this layer
+	}
+
+	@Override
+	public void sendRequest(Exchange exchange, Request request) {
+		if (CANCEL.equals(request.getOptions().getObserve())) {
+			/* TODO: don't send, if connection is not available */
+		}
+		super.sendRequest(exchange, request);
+	}
+
+	@Override
+	public void sendResponse(final Exchange exchange, Response response) {
+		final ObserveRelation relation = exchange.getRelation();
+		if (relation != null && relation.isEstablished()) {
+			if (!response.getOptions().hasObserve()) {
+				relation.cancel();
+				response.setLast(true);
+			} else {
+				response.setLast(false);
+			}
+		} // else no observe was requested or the resource does not allow it
+		super.sendResponse(exchange, response);
+	}
+
+	@Override
+	public void receiveResponse(Exchange exchange, Response response) {
+		if (response.getOptions().hasObserve() && exchange.getRequest().isCanceled()) {
```

Comment:
This is considered erroneous behavior of the server, right? According to A.3 of coap-over-tcp the client (this side) MUST actively cancel the observation by means of a GET request. This is what would have had to happen when the Request has been canceled, right?